containers/ftp/s3_files.rb

```diff
@@ -0,0 +1,103 @@
+require 'aws-sdk-s3'
+require 'digest/sha2'
+require 'tempfile'
+require 'fileutils'
+require 'logger'
+load './utils.rb'
+
+class S3Files
+  include Utils
+  attr_reader :s3, :logger, :retry_count
+
+  def initialize(client = nil, logger = nil, retry_count = 10)
+    @s3 = client || Aws::S3::Client.new
+    @logger = logger || Logger.new(STDOUT)
+    @retry_count = retry_count || ENV['S3_RETRY_COUNT']
+  end
+
+  def download_file(bucket, key)
+    file_info = nil
+    file_downloaded = false
+    temp_file = nil
+
+    tries = 0
+    while !file_info && tries < retry_count
+      begin
+        tries += 1
+        file_info = s3.head_object(bucket: bucket, key: key)
```

Comment:
We saw timing errors where a file was saved to s3, but didn't show up yet on `get`, so we added a small retry/timeout to wait for it before fully failing the task.

Reply:
this also allows us to get the content-length, so that can be used later for determining the file has been fully transfered.

Reply:
(n.b. maybe we can also check the md5 value locally of the file, to make sure the s3 download was not only complete, but correct...)

Reply:
added to the code comments